numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -478,92 +483,83 @@ NPY_NO_EXPORT int
 PyUFunc_SimpleUniformOperationTypeResolver(
         PyUFuncObject *ufunc,
         NPY_CASTING casting,
-        PyArrayObject **operands,
-        PyObject *type_tup,
+        PyArray_Descr **op_dtypes,
         PyArray_Descr **out_dtypes)
 {
-    const char *ufunc_name = ufunc_get_name_cstr(ufunc);
-
     if (ufunc->nin < 1) {
+        const char *ufunc_name = ufunc_get_name_cstr(ufunc);
+
         PyErr_Format(PyExc_RuntimeError, "ufunc %s is configured "
                 "to use uniform operation type resolution but has "
                 "no inputs",
                 ufunc_name);
         return -1;
     }
-    int nop = ufunc->nin + ufunc->nout;
+    int nin = ufunc->nin, nop = nin + ufunc->nout;
 
     /*
      * There's a custom data type or an object array
      */
     bool has_custom_or_object = false;
     for (int iop = 0; iop < ufunc->nin; iop++) {
-        int type_num = PyArray_DESCR(operands[iop])->type_num;
+        int type_num = op_dtypes[iop]->type_num;
         if (type_num >= NPY_NTYPES || type_num == NPY_OBJECT) {
             has_custom_or_object = true;
             break;
         }
     }
 
     if (has_custom_or_object) {
-        return PyUFunc_DefaultTypeResolver(ufunc, casting, operands,
-                type_tup, out_dtypes);
+        return PyUFunc_DefaultTypeResolver(ufunc, casting, op_dtypes, out_dtypes);
+    }
+
+    // TODO: I changed signature to fill in the result dtype, instead of first
+    //       so check result here as well. In principle this should detect
+    //       errors in the provided signature!
+    if (out_dtypes[nin] != NULL) {
+        if (out_dtypes[0] == NULL) {
+            Py_INCREF(out_dtypes[nin]);
+            out_dtypes[0] = out_dtypes[nin];
+        }
     }
 
-    if (type_tup == NULL) {
-        /* PyArray_ResultType forgets to force a byte order when n == 1 */
+    if (out_dtypes[0] == NULL) {
+        /* promotion does not force native byte order when n == 1 */
         if (ufunc->nin == 1){
-            out_dtypes[0] = ensure_dtype_nbo(PyArray_DESCR(operands[0]));
+            out_dtypes[0] = ensure_dtype_nbo(op_dtypes[0]);
         }
         else {
-            out_dtypes[0] = PyArray_ResultType(ufunc->nin, operands, 0, NULL);
+            // TODO: SHOULD USE PromoteTypeSequence, but crashes for me
+            //       for no apparent reason (runs fine in valgrind even)...
+            // out_dtypes[0] = PyArray_PromoteTypeSequence(op_dtypes, ufunc->nin);
+            if (ufunc->nin == 2) {
+                out_dtypes[0] = PyArray_PromoteTypes(op_dtypes[0], op_dtypes[1]);
+            }
+            else {
+                out_dtypes[0] = PyArray_ResultType(0, NULL, ufunc->nin, op_dtypes);
```

Comment:
If anyone can explain to me what can possibly cause that `PyArray_PromoteTypeSequence(...)` prints out the correct dtype before returning, and next thing `out_dtypes[0]` is corrupted, I would be very interested. (Maybe short of a compiler bug). As commented, it not crash inside valgrind...

Reply:
Probably becausethe signature of `PyArray_PromoteTypeSequence` is not known to this file, as it doesn't appear in any header. @tylerjereddy found a similar thing in one of my PRs, altough I though he added a compiler flag in CI to stop that happening silently again.

Reply:
Ah, now that makes sense. Seriously puzzled me, well, I did not run it through CI, so that compiler flag probably exists in CI, but not in `runtests.py`.